src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -218,6 +218,7 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
                 {
                     AllocateReadTail();
 
+                    Debug.Assert(_readTail != null);
```

Comment:
I would just use `!` on `_readTail`.  We just called `AllocateReadTail()` on the line above.